# 9 データベースの応用（高度なSQL処理と高速化）

- **[9.1 この章の概要](#9.1-この章の概要)**
    - [9.1.1 この章について](#9.1.1-この章について)
    - [9.1.2 高度なSQL処理](#9.1.2-高度なSQL処理)
    - [9.1.3 ストアドプロシージャとストアドファンクション](#9.1.3-ストアドプロシージャとストアドファンクション)
    - [9.1.4 SQLの高速化について](#9.1.4-SQLの高速化について)
    - [9.1.5 その他](#9.1.5-その他)
<br><br>
- **[9.2 高度なSQL処理](#9.2-高度なSQL処理)** 
    - [9.2.1 データの準備と確認](#9.2.1-データの準備と確認)
    - [9.2.2 exists文、case文、自己結合の応用](#9.2.2-exists文、case文、自己結合の応用)
    - [9.2.3 外部結合の応用](#9.2.3-外部結合の応用)
    - [9.2.4（応用）having句の応用](#9.2.4-（応用）having句の応用)
<br><br> 
- **[9.3 ストアドプロシージャとストアドファンクション](#9.3-ストアドプロシージャとスドアドファンクション)** 
    - [9.3.1 ストアドプロシージャ](#9.3.1-ストアドプロシージャ)
    - [9.3.2 ストアドファンクション](#9.3.2-ストアドファンクション)
<br><br>
- **[9.4 インデクシングとSQLの高速化](#9.4-インデクシングとSQLの高速化)**
    - [9.4.1 インデックスの設定](#9.4.1-インデックスの設定)
    - [9.4.2 インデックスの種類](#9.4.2-インデックスの種類)
    - [9.4.3 SQLのパフォーマンスチューニング](#9.4.3-SQLのパフォーマンスチューニング)
<br><br>
- **[9.5 その他](#9.5-その他)**
    - [9.5.1 データベースとファイルのやり取り](#9.5.1-データベースとファイルのやり取り)
    - [9.5.2 データモデリングと正規化](#9.5.2-データモデリングと正規化)
<br><br>
- **[9.6 総合問題](#9.6-総合問題)**
    - [9.6.1 総合問題1](#9.6.1-総合問題1)
    - [9.6.2 総合問題2](#9.6.2-総合問題2)

***

## 9.1 この章の概要
ゴール：複雑なSQL処理や高速化ができることを知る

### 9.1.1 この章について
キーワード：特になし

この章では、前章のデータベース基礎知識をベースに、さらにSQLの操作方法に慣れるため、また高度な処理ができるようになるための技法を身につけていただきます。ここをクリアできれば、SQLの中級者の道へと進み、後は独学でも学ぶことができるでしょう。以下の文献が参考になりますので、ここが終わった後は是非読むことをおすすめします。最後の参考文献『ビッグデータ分析・活用のためのSQLレシピ』は、PostgreSQLやRedshift等いろいろな記述方法が載っています。またマーケティング分析でよく使うテクニック等もあり、調べ物として便利です。

**注意：この章で使うデータベースやテーブルは前章で作成したもの（練習問題含む）も一部使いますので、前章のSQLは実行済みで、該当のデータがあるとしてこの章を進めていきます。**

>[参考文献]

>『SQL実践入門──高速でわかりやすいクエリの書き方 (WEB+DB PRESS plus) 』（ミック  (著)、技術評論社）

>『プログラマのためのSQL 第4版』（ジョー・セルコ (著), Joe Celko (著), ミック (監修, 翻訳)、翔泳社）

>『達人に学ぶ SQL徹底指南書』（ミック  (著)、翔泳社）

>『達人に学ぶDB設計 徹底指南書 初級者で終わりたくないあなたへ 』（ミック  (著)、翔泳社）

>『ビッグデータ分析・活用のためのSQLレシピ』（加嵜 長門  (著), 田宮 直人  (著), 丸山 弘詩 (編集)、マイナビ出版 ）

### 9.1.2 高度なSQL処理
キーワード：exists,自己結合,case文活用

前章までは短いSQL文を学んできましたが、この章では少し難しいSQLの処理(exists,差分などテクニカルな処理)や、比較的大きなデータや複数のデータソースを使って、複雑なSQL文を書く処理を学んでいきましょう。

### 9.1.3 ストアドプロシージャとストアドファンクション
キーワード：ストアドプロシージャ、ストアドファンクション

一連の処理をするストアドプロシージャについて学びます。SQL文を単発で使うのではなく、いくつか組み合わせたり、関数のように用いる（ストアドファンクション）こともできます。ここでは、それらの処理について学びましょう。

### 9.1.4 SQLの高速化について
キーワード：インデクシング

他のプログラミング言語同様、同じデータ抽出結果でもSQLの書き方は様々なパターンがあります。SQLの書き方によって、データを抽出するスピードが上がりますし、インデクシングというパフォーマンスチューニングによって、効率を上げることができます。

### 9.1.5 その他
キーワード：ファイルのやり取り、データモデリング、正規化

その他、データベースとSQLについて、知っておくと便利なこと、概念等について説明します。主に、csvファイルなどのやり取りとデータモデリング、正規化について学びます。

***

## 9.2 高度なSQL処理
ゴール：複雑な処理をSQLを使って計算することができる

### 9.2.1 データの準備と確認
キーワード：show,use,desc

前章までは短いSQL文を学んできましたが、この章では少し難しい処理をSQLで書いたり、比較的大きなデータや複数のデータソースを使います。ここで、複雑なSQL文を書く練習をしていきましょう。

この章では以下のデータを使います。このデータはある会社の従業員データで、従業員の属性や所属部署、給与等のデータが入っているデータベースです。まず、以前の章の総合問題で準備した同じ手順でデータを取得し、データベースに挿入します。（注意事項等を読んでください。）

https://launchpad.net/test-db/employees-db-1

terminal(もしくは画面上にあるKernelをPythonに変更して)から以下のコマンドを入力します。

-------------------
wget https://launchpad.net/test-db/employees-db-1/1.0.6/+download/employees_db-full-1.0.6.tar.bz2  
tar jxfv  employees_db-full-1.0.6.tar.bz2  
cd employees_db  
mysql -h"zdb" -P"3306" -u root -p"gci" < employees.sql  

-------------------

そして、employeesを選択してみます。次のSQLコマンドを実行してみましょう。

In [1]:
show databases;
use employees;

Database 
 
 
 
 
 0 
 TEST1 
 
 
 1 
 TEST2 
 
 
 2 
 employees 
 
 
 3 
 information_schema 
 
 
 4 
 mysql 
 
 
 5 
 performance_schema 
 
 
 6 
 world 
 
 
 Query OK, 0 rows affected.

テーブルを見てみましょう。テーブルが6つあるのが確認できます。

In [2]:
show tables;

Tables_in_employees 
 
 
 
 
 0 
 departments 
 
 
 1 
 dept_emp 
 
 
 2 
 dept_manager 
 
 
 3 
 employees 
 
 
 4 
 salaries 
 
 
 5 
 titles

それぞれのデータ構造を簡単に見てみます。従業員のデータで、部署テーブル、部署のマネージャーテーブル、給料テーブル、職種タイトルテーブルなどがあります。なお、salariesテーブルなどはレコード数が少し多い（100万レコード以上）ので、呼び出すときは注意しましょう。

In [3]:
desc departments;desc dept_emp;desc dept_manager;desc employees;desc salaries;desc titles;

Field 
 Type 
 Null 
 Key 
 Default 
 Extra 
 
 
 
 
 0 
 dept_no 
 char(4) 
 NO 
 PRI 
 None 
 
 
 
 1 
 dept_name 
 varchar(40) 
 NO 
 UNI 
 None 
 
 
 
 
 
 
 
 Field 
 Type 
 Null 
 Key 
 Default 
 Extra 
 
 
 
 
 0 
 emp_no 
 int(11) 
 NO 
 PRI 
 None 
 
 
 
 1 
 dept_no 
 char(4) 
 NO 
 PRI 
 None 
 
 
 
 2 
 from_date 
 date 
 NO 
 
 None 
 
 
 
 3 
 to_date 
 date 
 NO 
 
 None 
 
 
 
 
 
 
 
 Field 
 Type 
 Null 
 Key 
 Default 
 Extra 
 
 
 
 
 0 
 dept_no 
 char(4) 
 NO 
 PRI 
 None 
 
 
 
 1 
 emp_no 
 int(11) 
 NO 
 PRI 
 None 
 
 
 
 2 
 from_date 
 date 
 NO 
 
 None 
 
 
 
 3 
 to_date 
 date 
 NO 
 
 None 
 
 
 
 
 
 
 
 Field 
 Type 
 Null 
 Key 
 Default 
 Extra 
 
 
 
 
 0 
 emp_no 
 int(11) 
 NO 
 PRI 
 None 
 
 
 
 1 
 birth_date 
 date 
 NO 
 
 None 
 
 
 
 2 
 first_name 
 varchar(14) 
 NO 
 
 None 
 
 
 
 3 
 last_name 
 varchar(16) 
 NO 
 
 None 
 
 
 
 4 
 gender 
 enum('M','F') 
 NO 
 
 None 
 
 
 
 5 
 hire_date 
 date 
 NO 
 
 None 
 
 
 
 
 
 
 
 Field 
 Type 
 Null 
 Key 
 Default 
 Extra 
 
 
 
 
 0 
 emp_no 
 int(11) 
 NO 
 PRI 
 None 
 
 
 
 1 
 salary 
 int(11) 
 NO 
 
 None 
 
 
 
 2 
 from_date 
 date 
 NO 
 PRI 
 None 
 
 
 
 3 
 to_date 
 date 
 NO 
 
 None 
 
 
 
 
 
 
 
 Field 
 Type 
 Null 
 Key 
 Default 
 Extra 
 
 
 
 
 0 
 emp_no 
 int(11) 
 NO 
 PRI 
 None 
 
 
 
 1 
 title 
 varchar(50) 
 NO 
 PRI 
 None 
 
 
 
 2 
 from_date 
 date 
 NO 
 PRI 
 None 
 
 
 
 3 
 to_date 
 date 
 YES 
 
 None

以下は参考ですが、ER図といい、それぞれのテーブルの関係性を示した図です。テーブルを設計をしたり、テーブル間の関係を確認するときによく使います。

![comment](https://dev.mysql.com/doc/employee/en/images/employees-schema.png)

参照URL：https://dev.mysql.com/doc/employee/en/sakila-structure.html

### 9.2.2 exists文、case文、自己結合の応用
キーワード：条件分岐、自己結合

ここでは前章では取り扱わなかったexists文や自己結合を使った処理を学びます。

exists文は、以下のように記述し、副問合せによって返されたレコードが一つでもあれば真，一つもなければ偽を返します。

In [4]:
select * from dept_manager a
where exists ( select * from departments b where a.dept_no = b.dept_no) limit 10;

dept_no 
 emp_no 
 from_date 
 to_date 
 
 
 
 
 0 
 d001 
 110022 
 1985-01-01 
 1991-10-01 
 
 
 1 
 d001 
 110039 
 1991-10-01 
 9999-01-01 
 
 
 2 
 d002 
 110085 
 1985-01-01 
 1989-12-17 
 
 
 3 
 d002 
 110114 
 1989-12-17 
 9999-01-01 
 
 
 4 
 d003 
 110183 
 1985-01-01 
 1992-03-21 
 
 
 5 
 d003 
 110228 
 1992-03-21 
 9999-01-01 
 
 
 6 
 d004 
 110303 
 1985-01-01 
 1988-09-09 
 
 
 7 
 d004 
 110344 
 1988-09-09 
 1992-08-02 
 
 
 8 
 d004 
 110386 
 1992-08-02 
 1996-08-30 
 
 
 9 
 d004 
 110420 
 1996-08-30 
 9999-01-01

ちなみに上のexists文での操作は以下のクエリでも同じ操作ができます。exists文の方が処理が速いことがあるので、exists文は覚えておくとよいでしょう。

In [5]:
select a.* from dept_manager a, departments b
where a.dept_no = b.dept_no limit 10;

dept_no 
 emp_no 
 from_date 
 to_date 
 
 
 
 
 0 
 d001 
 110022 
 1985-01-01 
 1991-10-01 
 
 
 1 
 d001 
 110039 
 1991-10-01 
 9999-01-01 
 
 
 2 
 d002 
 110085 
 1985-01-01 
 1989-12-17 
 
 
 3 
 d002 
 110114 
 1989-12-17 
 9999-01-01 
 
 
 4 
 d003 
 110183 
 1985-01-01 
 1992-03-21 
 
 
 5 
 d003 
 110228 
 1992-03-21 
 9999-01-01 
 
 
 6 
 d004 
 110303 
 1985-01-01 
 1988-09-09 
 
 
 7 
 d004 
 110344 
 1988-09-09 
 1992-08-02 
 
 
 8 
 d004 
 110386 
 1992-08-02 
 1996-08-30 
 
 
 9 
 d004 
 110420 
 1996-08-30 
 9999-01-01

次は、case文を活用した例です。job titleを見てみますと7種類あるのがわかります。ここで、このタイトルを新しいカテゴリーに分類して、それぞれのレコード数をカウントしてみましょう。現在のjob title毎のレコード数は、以下のようになります。（なお、1人で2つ以上のタイトルが付いている人がいますが、今は特に気にしないことにします。）

In [6]:
select title, count(*) 
from titles 
group by title;

title 
 count(*) 
 
 
 
 
 0 
 Assistant Engineer 
 15128 
 
 
 1 
 Engineer 
 115003 
 
 
 2 
 Manager 
 24 
 
 
 3 
 Senior Engineer 
 97750 
 
 
 4 
 Senior Staff 
 92853 
 
 
 5 
 Staff 
 107391 
 
 
 6 
 Technique Leader 
 15159

ここで、これらのタイトルを新しいカテゴリーに分類して、それぞれのレコード数をカウントしてみましょう。具体的には、Engineerとついている人はEngineer、Staffとついている人はStaff、LeaderかManagerとついている人はLeaderで大分類としてカテゴリーを分けて、それぞれのレコード数をカウントします。

これは、何かのカテゴリーや分類があった時に、もう少し大きな分類で集計したいときなどに使うテクニックです。例えば、データとして都道府県があった場合、各都道府県で集計すると数が多くなるので、関東や関西など、ある程度大きな分類で考えたい時に使います。

In [7]:
select case title
        when 'Engineer' then 'Engineer'
        when 'Staff' then 'Staff'
        when 'Senior Engineer' then 'Engineer'
        when 'Senior Staff ' then 'Staff'
        when 'Technique Leader' then 'Leader'
        when 'Assistant Engineer' then 'Engineer'
        when 'Manager' then 'Leader'
    else 'others' end as category
    ,count(*) as cnt
from titles
group by category;

category 
 cnt 
 
 
 
 
 0 
 Engineer 
 227881 
 
 
 1 
 Leader 
 15183 
 
 
 2 
 Staff 
 200244

次は、自分自身のテーブルと結合する**自己結合**です。情報としては自分のテーブルだけあれば良いように思えますが、どういった場面で自己結合が役に立つのか、それを見ていくことにします。

以下は、depatmentsテーブルを使って集計例を見てみます。部署は9種類あるのがわかります。

In [8]:
select * from departments order by 1;

dept_no 
 dept_name 
 
 
 
 
 0 
 d001 
 Marketing 
 
 
 1 
 d002 
 Finance 
 
 
 2 
 d003 
 Human Resources 
 
 
 3 
 d004 
 Production 
 
 
 4 
 d005 
 Development 
 
 
 5 
 d006 
 Quality Management 
 
 
 6 
 d007 
 Sales 
 
 
 7 
 d008 
 Research 
 
 
 8 
 d009 
 Customer Service

データのレコードから、すべての組み合わせ（重複ありで自身の組み合わせ含む）のテーブルを作りたいときがある場合に、以下のようにして、自己結合します。9種類あったので、81通り（9×9）あります。（実際に81通りであることを確認するため、ここではlimit 100と末尾に加えることでデフォルトの表示行数である30行よりも多いレコードを取得しています。以下、同様です。）

In [9]:
select a.*,b.* from departments a, departments b limit 100;

dept_no 
 dept_name 
 dept_no 
 dept_name 
 
 
 
 
 0 
 d009 
 Customer Service 
 d001 
 Marketing 
 
 
 1 
 d005 
 Development 
 d001 
 Marketing 
 
 
 2 
 d002 
 Finance 
 d001 
 Marketing 
 
 
 3 
 d003 
 Human Resources 
 d001 
 Marketing 
 
 
 4 
 d001 
 Marketing 
 d001 
 Marketing 
 
 
 5 
 d004 
 Production 
 d001 
 Marketing 
 
 
 6 
 d006 
 Quality Management 
 d001 
 Marketing 
 
 
 7 
 d008 
 Research 
 d001 
 Marketing 
 
 
 8 
 d007 
 Sales 
 d001 
 Marketing 
 
 
 9 
 d009 
 Customer Service 
 d002 
 Finance 
 
 
 10 
 d005 
 Development 
 d002 
 Finance 
 
 
 11 
 d002 
 Finance 
 d002 
 Finance 
 
 
 12 
 d003 
 Human Resources 
 d002 
 Finance 
 
 
 13 
 d001 
 Marketing 
 d002 
 Finance 
 
 
 14 
 d004 
 Production 
 d002 
 Finance 
 
 
 15 
 d006 
 Quality Management 
 d002 
 Finance 
 
 
 16 
 d008 
 Research 
 d002 
 Finance 
 
 
 17 
 d007 
 Sales 
 d002 
 Finance 
 
 
 18 
 d009 
 Customer Service 
 d003 
 Human Resources 
 
 
 19 
 d005 
 Development 
 d003 
 Human Resources 
 
 
 20 
 d002 
 Finance 
 d003 
 Human Resources 
 
 
 21 
 d003 
 Human Resources 
 d003 
 Human Resources 
 
 
 22 
 d001 
 Marketing 
 d003 
 Human Resources 
 
 
 23 
 d004 
 Production 
 d003 
 Human Resources 
 
 
 24 
 d006 
 Quality Management 
 d003 
 Human Resources 
 
 
 25 
 d008 
 Research 
 d003 
 Human Resources 
 
 
 26 
 d007 
 Sales 
 d003 
 Human Resources 
 
 
 27 
 d009 
 Customer Service 
 d004 
 Production 
 
 
 28 
 d005 
 Development 
 d004 
 Production 
 
 
 29 
 d002 
 Finance 
 d004 
 Production 
 
 
 30 
 d003 
 Human Resources 
 d004 
 Production 
 
 
 31 
 d001 
 Marketing 
 d004 
 Production 
 
 
 32 
 d004 
 Production 
 d004 
 Production 
 
 
 33 
 d006 
 Quality Management 
 d004 
 Production 
 
 
 34 
 d008 
 Research 
 d004 
 Production 
 
 
 35 
 d007 
 Sales 
 d004 
 Production 
 
 
 36 
 d009 
 Customer Service 
 d005 
 Development 
 
 
 37 
 d005 
 Development 
 d005 
 Development 
 
 
 38 
 d002 
 Finance 
 d005 
 Development 
 
 
 39 
 d003 
 Human Resources 
 d005 
 Development 
 
 
 40 
 d001 
 Marketing 
 d005 
 Development 
 
 
 41 
 d004 
 Production 
 d005 
 Development 
 
 
 42 
 d006 
 Quality Management 
 d005 
 Development 
 
 
 43 
 d008 
 Research 
 d005 
 Development 
 
 
 44 
 d007 
 Sales 
 d005 
 Development 
 
 
 45 
 d009 
 Customer Service 
 d006 
 Quality Management 
 
 
 46 
 d005 
 Development 
 d006 
 Quality Management 
 
 
 47 
 d002 
 Finance 
 d006 
 Quality Management 
 
 
 48 
 d003 
 Human Resources 
 d006 
 Quality Management 
 
 
 49 
 d001 
 Marketing 
 d006 
 Quality Management 
 
 
 50 
 d004 
 Production 
 d006 
 Quality Management 
 
 
 51 
 d006 
 Quality Management 
 d006 
 Quality Management 
 
 
 52 
 d008 
 Research 
 d006 
 Quality Management 
 
 
 53 
 d007 
 Sales 
 d006 
 Quality Management 
 
 
 54 
 d009 
 Customer Service 
 d007 
 Sales 
 
 
 55 
 d005 
 Development 
 d007 
 Sales 
 
 
 56 
 d002 
 Finance 
 d007 
 Sales 
 
 
 57 
 d003 
 Human Resources 
 d007 
 Sales 
 
 
 58 
 d001 
 Marketing 
 d007 
 Sales 
 
 
 59 
 d004 
 Production 
 d007 
 Sales 
 
 
 60 
 d006 
 Quality Management 
 d007 
 Sales 
 
 
 61 
 d008 
 Research 
 d007 
 Sales 
 
 
 62 
 d007 
 Sales 
 d007 
 Sales 
 
 
 63 
 d009 
 Customer Service 
 d008 
 Research 
 
 
 64 
 d005 
 Development 
 d008 
 Research 
 
 
 65 
 d002 
 Finance 
 d008 
 Research 
 
 
 66 
 d003 
 Human Resources 
 d008 
 Research 
 
 
 67 
 d001 
 Marketing 
 d008 
 Research 
 
 
 68 
 d004 
 Production 
 d008 
 Research 
 
 
 69 
 d006 
 Quality Management 
 d008 
 Research 
 
 
 70 
 d008 
 Research 
 d008 
 Research 
 
 
 71 
 d007 
 Sales 
 d008 
 Research 
 
 
 72 
 d009 
 Customer Service 
 d009 
 Customer Service 
 
 
 73 
 d005 
 Development 
 d009 
 Customer Service 
 
 
 74 
 d002 
 Finance 
 d009 
 Customer Service 
 
 
 75 
 d003 
 Human Resources 
 d009 
 Customer Service 
 
 
 76 
 d001 
 Marketing 
 d009 
 Customer Service 
 
 
 77 
 d004 
 Production 
 d009 
 Customer Service

すべての組み合わせで、自身の組み合わせがないパターンを作成したい場合は、以下のように「a.dept_no != b.dept_no 」として、自分自身と一致するもの以外を抽出します。例えば、d001とd001の組み合わせやd009とd009の組み合わせ等は省かれます。自分自身と同じものは9パターンあるので、結果以下のクエリは81-9=72通りあります。

In [10]:
select a.*,b.* 
from departments a,departments b
where a.dept_no != b.dept_no limit 100;

dept_no 
 dept_name 
 dept_no 
 dept_name 
 
 
 
 
 0 
 d009 
 Customer Service 
 d001 
 Marketing 
 
 
 1 
 d005 
 Development 
 d001 
 Marketing 
 
 
 2 
 d002 
 Finance 
 d001 
 Marketing 
 
 
 3 
 d003 
 Human Resources 
 d001 
 Marketing 
 
 
 4 
 d004 
 Production 
 d001 
 Marketing 
 
 
 5 
 d006 
 Quality Management 
 d001 
 Marketing 
 
 
 6 
 d008 
 Research 
 d001 
 Marketing 
 
 
 7 
 d007 
 Sales 
 d001 
 Marketing 
 
 
 8 
 d009 
 Customer Service 
 d002 
 Finance 
 
 
 9 
 d005 
 Development 
 d002 
 Finance 
 
 
 10 
 d003 
 Human Resources 
 d002 
 Finance 
 
 
 11 
 d001 
 Marketing 
 d002 
 Finance 
 
 
 12 
 d004 
 Production 
 d002 
 Finance 
 
 
 13 
 d006 
 Quality Management 
 d002 
 Finance 
 
 
 14 
 d008 
 Research 
 d002 
 Finance 
 
 
 15 
 d007 
 Sales 
 d002 
 Finance 
 
 
 16 
 d009 
 Customer Service 
 d003 
 Human Resources 
 
 
 17 
 d005 
 Development 
 d003 
 Human Resources 
 
 
 18 
 d002 
 Finance 
 d003 
 Human Resources 
 
 
 19 
 d001 
 Marketing 
 d003 
 Human Resources 
 
 
 20 
 d004 
 Production 
 d003 
 Human Resources 
 
 
 21 
 d006 
 Quality Management 
 d003 
 Human Resources 
 
 
 22 
 d008 
 Research 
 d003 
 Human Resources 
 
 
 23 
 d007 
 Sales 
 d003 
 Human Resources 
 
 
 24 
 d009 
 Customer Service 
 d004 
 Production 
 
 
 25 
 d005 
 Development 
 d004 
 Production 
 
 
 26 
 d002 
 Finance 
 d004 
 Production 
 
 
 27 
 d003 
 Human Resources 
 d004 
 Production 
 
 
 28 
 d001 
 Marketing 
 d004 
 Production 
 
 
 29 
 d006 
 Quality Management 
 d004 
 Production 
 
 
 30 
 d008 
 Research 
 d004 
 Production 
 
 
 31 
 d007 
 Sales 
 d004 
 Production 
 
 
 32 
 d009 
 Customer Service 
 d005 
 Development 
 
 
 33 
 d002 
 Finance 
 d005 
 Development 
 
 
 34 
 d003 
 Human Resources 
 d005 
 Development 
 
 
 35 
 d001 
 Marketing 
 d005 
 Development 
 
 
 36 
 d004 
 Production 
 d005 
 Development 
 
 
 37 
 d006 
 Quality Management 
 d005 
 Development 
 
 
 38 
 d008 
 Research 
 d005 
 Development 
 
 
 39 
 d007 
 Sales 
 d005 
 Development 
 
 
 40 
 d009 
 Customer Service 
 d006 
 Quality Management 
 
 
 41 
 d005 
 Development 
 d006 
 Quality Management 
 
 
 42 
 d002 
 Finance 
 d006 
 Quality Management 
 
 
 43 
 d003 
 Human Resources 
 d006 
 Quality Management 
 
 
 44 
 d001 
 Marketing 
 d006 
 Quality Management 
 
 
 45 
 d004 
 Production 
 d006 
 Quality Management 
 
 
 46 
 d008 
 Research 
 d006 
 Quality Management 
 
 
 47 
 d007 
 Sales 
 d006 
 Quality Management 
 
 
 48 
 d009 
 Customer Service 
 d007 
 Sales 
 
 
 49 
 d005 
 Development 
 d007 
 Sales 
 
 
 50 
 d002 
 Finance 
 d007 
 Sales 
 
 
 51 
 d003 
 Human Resources 
 d007 
 Sales 
 
 
 52 
 d001 
 Marketing 
 d007 
 Sales 
 
 
 53 
 d004 
 Production 
 d007 
 Sales 
 
 
 54 
 d006 
 Quality Management 
 d007 
 Sales 
 
 
 55 
 d008 
 Research 
 d007 
 Sales 
 
 
 56 
 d009 
 Customer Service 
 d008 
 Research 
 
 
 57 
 d005 
 Development 
 d008 
 Research 
 
 
 58 
 d002 
 Finance 
 d008 
 Research 
 
 
 59 
 d003 
 Human Resources 
 d008 
 Research 
 
 
 60 
 d001 
 Marketing 
 d008 
 Research 
 
 
 61 
 d004 
 Production 
 d008 
 Research 
 
 
 62 
 d006 
 Quality Management 
 d008 
 Research 
 
 
 63 
 d007 
 Sales 
 d008 
 Research 
 
 
 64 
 d005 
 Development 
 d009 
 Customer Service 
 
 
 65 
 d002 
 Finance 
 d009 
 Customer Service 
 
 
 66 
 d003 
 Human Resources 
 d009 
 Customer Service 
 
 
 67 
 d001 
 Marketing 
 d009 
 Customer Service 
 
 
 68 
 d004 
 Production 
 d009 
 Customer Service 
 
 
 69 
 d006 
 Quality Management 
 d009 
 Customer Service 
 
 
 70 
 d008 
 Research 
 d009 
 Customer Service 
 
 
 71 
 d007 
 Sales 
 d009 
 Customer Service

先ほどは、「d009とd001」と「d001とd009」のように、実質的に重複するレコードが含まれますが、重複がなしですべての組み合わせを抽出したい場合は、以下のようにします。9種類から2種類選ぶので、${}_9 C _2=36$通りで、36行表示されます。

In [11]:
select a.*,b.* 
from departments a,departments b
where a.dept_no > b.dept_no limit 100;

dept_no 
 dept_name 
 dept_no 
 dept_name 
 
 
 
 
 0 
 d009 
 Customer Service 
 d001 
 Marketing 
 
 
 1 
 d009 
 Customer Service 
 d002 
 Finance 
 
 
 2 
 d009 
 Customer Service 
 d003 
 Human Resources 
 
 
 3 
 d009 
 Customer Service 
 d004 
 Production 
 
 
 4 
 d009 
 Customer Service 
 d005 
 Development 
 
 
 5 
 d009 
 Customer Service 
 d006 
 Quality Management 
 
 
 6 
 d009 
 Customer Service 
 d007 
 Sales 
 
 
 7 
 d009 
 Customer Service 
 d008 
 Research 
 
 
 8 
 d005 
 Development 
 d001 
 Marketing 
 
 
 9 
 d005 
 Development 
 d002 
 Finance 
 
 
 10 
 d005 
 Development 
 d003 
 Human Resources 
 
 
 11 
 d005 
 Development 
 d004 
 Production 
 
 
 12 
 d002 
 Finance 
 d001 
 Marketing 
 
 
 13 
 d003 
 Human Resources 
 d001 
 Marketing 
 
 
 14 
 d003 
 Human Resources 
 d002 
 Finance 
 
 
 15 
 d004 
 Production 
 d001 
 Marketing 
 
 
 16 
 d004 
 Production 
 d002 
 Finance 
 
 
 17 
 d004 
 Production 
 d003 
 Human Resources 
 
 
 18 
 d006 
 Quality Management 
 d001 
 Marketing 
 
 
 19 
 d006 
 Quality Management 
 d002 
 Finance 
 
 
 20 
 d006 
 Quality Management 
 d003 
 Human Resources 
 
 
 21 
 d006 
 Quality Management 
 d004 
 Production 
 
 
 22 
 d006 
 Quality Management 
 d005 
 Development 
 
 
 23 
 d008 
 Research 
 d001 
 Marketing 
 
 
 24 
 d008 
 Research 
 d002 
 Finance 
 
 
 25 
 d008 
 Research 
 d003 
 Human Resources 
 
 
 26 
 d008 
 Research 
 d004 
 Production 
 
 
 27 
 d008 
 Research 
 d005 
 Development 
 
 
 28 
 d008 
 Research 
 d006 
 Quality Management 
 
 
 29 
 d008 
 Research 
 d007 
 Sales 
 
 
 30 
 d007 
 Sales 
 d001 
 Marketing 
 
 
 31 
 d007 
 Sales 
 d002 
 Finance 
 
 
 32 
 d007 
 Sales 
 d003 
 Human Resources 
 
 
 33 
 d007 
 Sales 
 d004 
 Production 
 
 
 34 
 d007 
 Sales 
 d005 
 Development 
 
 
 35 
 d007 
 Sales 
 d006 
 Quality Management

さて、この自己結合、一見すると有用性を感じないのですが、これを活用することで、例えば、ある項目のデータのランキング結果を表示させることが可能です。以下、salariesテーブルの中から、従業員（emp_no）の給料が高い人を降順に並べ、ランキングを表示する列を追加する処理をしています。このような集計をしたい場合に、自己結合が使えます。

In [12]:
select a.emp_no
    ,a.salary
    ,a.from_date
    ,a.to_date
    ,(
        select count(b.salary)
        from salaries b
        where b.salary > a.salary
    ) + 1 as rank_1
from salaries a order by a.salary desc limit 10;

emp_no 
 salary 
 from_date 
 to_date 
 rank_1 
 
 
 
 
 0 
 43624 
 158220 
 2002-03-22 
 9999-01-01 
 1 
 
 
 1 
 43624 
 157821 
 2001-03-22 
 2002-03-22 
 2 
 
 
 2 
 254466 
 156286 
 2001-08-04 
 9999-01-01 
 3 
 
 
 3 
 47978 
 155709 
 2002-07-14 
 9999-01-01 
 4 
 
 
 4 
 253939 
 155513 
 2002-04-11 
 9999-01-01 
 5 
 
 
 5 
 109334 
 155377 
 2000-02-12 
 2001-02-11 
 6 
 
 
 6 
 109334 
 155190 
 2002-02-11 
 9999-01-01 
 7 
 
 
 7 
 109334 
 154888 
 2001-02-11 
 2002-02-11 
 8 
 
 
 8 
 109334 
 154885 
 1999-02-12 
 2000-02-12 
 9 
 
 
 9 
 80823 
 154459 
 2002-02-22 
 9999-01-01 
 10

#### <練習問題 1>

case文を使って部署毎にgender別の人数を集計してみましょう。(特に部署移動等は考慮しなくて良いです。)

ヒント：case文をsum関数の中に記述できます。

#### <練習問題 2>

job titleの全ての組み合わせ（重複なし）を表示してみましょう。

#### <練習問題 3>

exists文を使って部署毎に現在job titleがStaffである人数を集計してみましょう。なお、現在とは、to_dateが9999-01-01のことを指すとします。

### 9.2.3 外部結合の応用
キーワード：クロス集計、差分集合

ここでは外部結合を活用してみることにしましょう。

SQLを使ったクロス集計を実施してみます。以下は、各従業員（emp_no）がどのjob titleを持っているか（いたか）をクロス表にしています。外部結合でNULLになった場合は該当しないため空白とし、それ以外は該当するため'OK'と表示しています。Staff、Senior Staff、Engineerを対象に集計しています。

In [13]:
select 
    x0.emp_no
    ,case when x1.emp_no is not null then 'OK' else '' end as "Staff"
    ,case when x2.emp_no is not null then 'OK' else '' end as "Senior Staff"
    ,case when x3.emp_no is not null then 'OK' else '' end as "Engineer"
from (select distinct emp_no from employees ) x0
left outer join (select emp_no from titles where title ="Staff") x1 on x0.emp_no = x1.emp_no
left outer join (select emp_no from titles where title ="Senior Staff") x2 on x0.emp_no = x2.emp_no
left outer join (select emp_no from titles where title ="Engineer") x3 on x0.emp_no = x3.emp_no
limit 10;

emp_no 
 Staff 
 Senior Staff 
 Engineer 
 
 
 
 
 0 
 10001 
 
 
 
 
 
 1 
 10002 
 OK 
 
 
 
 
 2 
 10003 
 
 
 
 
 
 3 
 10004 
 
 
 OK 
 
 
 4 
 10005 
 OK 
 OK 
 
 
 
 5 
 10006 
 
 
 
 
 
 6 
 10007 
 OK 
 OK 
 
 
 
 7 
 10008 
 
 
 
 
 
 8 
 10009 
 
 
 OK 
 
 
 9 
 10010 
 
 
 OK

例えば、emp_noが10001の人は3タイトルどれにもなったことがなく、emp_noが10005の人は、StaffとSenior Staff 両方なったことがあるということがわかります。

次は、外部結合を活用した、差分集合についてです。差分集合とは、ある集合Aと集合Bがあった場合に、Aの中からBである人を抜いた集合A-Bを取り出した集合になります。以下の参照図だと、左の丸がA、右の丸がBの集合だとすると、A-Bは赤い部分を指します。

![comment](https://upload.wikimedia.org/wikipedia/commons/thumb/e/e6/Venn0100.svg/220px-Venn0100.svg.png)

参照URL:https://upload.wikimedia.org/wikipedia/commons/thumb/e/e6/Venn0100.svg/220px-Venn0100.svg.png

job titleテーブルを使って、まずはそれぞれのjob titleで何人いるかカウントしています。

In [14]:
select title,count(emp_no) as cnt from titles group by 1;

title 
 cnt 
 
 
 
 
 0 
 Assistant Engineer 
 15128 
 
 
 1 
 Engineer 
 115003 
 
 
 2 
 Manager 
 24 
 
 
 3 
 Senior Engineer 
 97750 
 
 
 4 
 Senior Staff 
 92853 
 
 
 5 
 Staff 
 107391 
 
 
 6 
 Technique Leader 
 15159

Staff とSenior Staffに注目してください。上記の集計では、それぞれ107,391人、92,853人いました。以下の集計で、どちらともこのポジションになっている人がいることがわかります（107391 + 92853 > 133981）。

In [15]:
select count(distinct emp_no) as cnt from titles where title='Senior Staff' or title='Staff'

cnt 
 
 
 
 
 0 
 133981

それでは、Staffだけになった人を抽出（カウント）するにはどうすればよいでしょうか。集合として、$A$をStaffになったの集合、$B$をSenior Staffになった人の集合すると、$A-B$と差分集合を求めるSQLは以下になります。

In [16]:
select * from 
(select * from titles where title='Staff') a
left outer join
(select * from titles where title='Senior Staff') b
on a.emp_no = b.emp_no
where b.title is NULL
limit 10;

emp_no 
 title 
 from_date 
 to_date 
 emp_no 
 title 
 from_date 
 to_date 
 
 
 
 
 0 
 10002 
 Staff 
 1996-08-03 
 9999-01-01 
 None 
 None 
 None 
 None 
 
 
 1 
 10011 
 Staff 
 1990-01-22 
 1996-11-09 
 None 
 None 
 None 
 None 
 
 
 2 
 10016 
 Staff 
 1998-02-11 
 9999-01-01 
 None 
 None 
 None 
 None 
 
 
 3 
 10019 
 Staff 
 1999-04-30 
 9999-01-01 
 None 
 None 
 None 
 None 
 
 
 4 
 10034 
 Staff 
 1995-04-12 
 1999-10-31 
 None 
 None 
 None 
 None 
 
 
 5 
 10055 
 Staff 
 1992-04-27 
 1995-07-22 
 None 
 None 
 None 
 None 
 
 
 6 
 10064 
 Staff 
 1985-11-20 
 1992-03-02 
 None 
 None 
 None 
 None 
 
 
 7 
 10071 
 Staff 
 1995-08-05 
 9999-01-01 
 None 
 None 
 None 
 None 
 
 
 8 
 10077 
 Staff 
 1994-12-23 
 9999-01-01 
 None 
 None 
 None 
 None 
 
 
 9 
 10082 
 Staff 
 1990-01-03 
 1990-01-15 
 None 
 None 
 None 
 None

また、以下はSenior Staffだけになった人です。

In [17]:
select * from 
(select * from titles where title='Senior Staff') a
left outer join
(select * from titles where title='Staff') b
on a.emp_no = b.emp_no
where b.title is NULL
limit 10;

emp_no 
 title 
 from_date 
 to_date 
 emp_no 
 title 
 from_date 
 to_date 
 
 
 
 
 0 
 10013 
 Senior Staff 
 1985-10-20 
 9999-01-01 
 None 
 None 
 None 
 None 
 
 
 1 
 10015 
 Senior Staff 
 1992-09-19 
 1993-08-22 
 None 
 None 
 None 
 None 
 
 
 2 
 10036 
 Senior Staff 
 1992-04-28 
 9999-01-01 
 None 
 None 
 None 
 None 
 
 
 3 
 10053 
 Senior Staff 
 1994-11-13 
 9999-01-01 
 None 
 None 
 None 
 None 
 
 
 4 
 10058 
 Senior Staff 
 1988-04-25 
 9999-01-01 
 None 
 None 
 None 
 None 
 
 
 5 
 10059 
 Senior Staff 
 1991-06-26 
 9999-01-01 
 None 
 None 
 None 
 None 
 
 
 6 
 10061 
 Senior Staff 
 1989-12-02 
 9999-01-01 
 None 
 None 
 None 
 None 
 
 
 7 
 10094 
 Senior Staff 
 1987-04-18 
 1997-11-08 
 None 
 None 
 None 
 None 
 
 
 8 
 10104 
 Senior Staff 
 1987-04-16 
 9999-01-01 
 None 
 None 
 None 
 None 
 
 
 9 
 10107 
 Senior Staff 
 1999-03-30 
 9999-01-01 
 None 
 None 
 None 
 None

#### <練習問題 1>

各従業員について、最大給料が70,000以下の人、70,000～80,000の人、80,000～90,000の人、90,000より多い人でクロス表を作成してみましょう。

#### <練習問題 2>

最大の給料が100,000より多い人のうち、最小の給料が90,000以上の人の数を求めてみましょう。

### 9.2.4 （応用）having句の応用
キーワード：having句

havingの応用です。次は各部署に所属している従業員の平均給与を計算し、その平均給与が60000以上の人が半分（50%）以上の部署を抽出します。（本来は年月日によって所属部署や給与が異なるので、それらを考慮し条件を絞る必要がありますが、少し複雑になりますので、今回はこの手法を理解するために、省略します。）

In [18]:
select 
    c.dept_no
    ,c.dept_name
    ,count(distinct a.emp_no) as cnt
    ,avg(avgSalary) as avgDepSalary 
from dept_emp a
left outer join
(
    select 
        emp_no
        ,avg(salary) as avgSalary 
    from salaries 
    group by 1
) b
on a.emp_no = b.emp_no 
left outer join departments c
on a.dept_no = c.dept_no
group by 1
having count(*) * 0.5 <= SUM(case when b.avgSalary >= 60000 then 1 else 0 end);

dept_no 
 dept_name 
 cnt 
 avgDepSalary 
 
 
 
 
 0 
 d001 
 Marketing 
 20211 
 69541.61771136 
 
 
 1 
 d002 
 Finance 
 17346 
 68061.43501801 
 
 
 2 
 d007 
 Sales 
 52245 
 78313.22247361

***

## 9.3 ストアドプロシージャとストアドファンクション
ゴール：ストアドプロシージャとストアドファンクションについて操作できること

### 9.3.1 ストアドプロシージャ
キーワード：ストアドプロシージャ

SQLを使った一連の処理をまとめて実行するためのストアドプロシージャについて学んでいきましょう。前までは、SQLを単発で実行してきました。しかし、連続して同じような処理を実行したい場合など、その都度それらを実行するのは面倒です。そこで、それらの処理を一括にまとめて処理するストアドプロシージャを用意して実行すれば、作業を効率化できます。
また、変数を用いた演算、条件分岐やテーブルの作成など多少複雑な処理も可能になります。ここでは、簡単なストアドプロシージャの作成について学びます。

In [19]:
show databases;
use TEST1;

Database 
 
 
 
 
 0 
 TEST1 
 
 
 1 
 TEST2 
 
 
 2 
 employees 
 
 
 3 
 information_schema 
 
 
 4 
 mysql 
 
 
 5 
 performance_schema 
 
 
 6 
 world 
 
 
 Query OK, 0 rows affected.

次に、ストアドプロシージャ作成をするのは、create procedureで始まり、beginとendでsql文を書きます。

In [20]:
create procedure stpr1()
begin
select * from meibo;
select * from meibo2;
end

Query OK, 0 rows affected.

作成したストアドプロシージャを呼ぶときは、以下のcallを使って実行します。

In [21]:
call stpr1;

id 
 name 
 class 
 age 
 height 
 
 
 
 
 0 
 1 
 Yamada 
 1 
 14 
 150 
 
 
 1 
 2 
 Tanaka 
 2 
 13 
 150 
 
 
 2 
 3 
 Suzuki 
 1 
 13 
 150 
 
 
 3 
 4 
 Kato 
 2 
 15 
 150 
 
 
 4 
 5 
 Ito 
 3 
 12 
 155 
 
 
 5 
 6 
 Takeuchi 
 2 
 16 
 155 
 
 
 6 
 7 
 Kimura 
 3 
 11 
 160 
 
 
 7 
 8 
 Sato 
 1 
 14 
 160 
 
 
 8 
 100 
 dummy_Yamada 
 None 
 None 
 None 
 
 
 9 
 200 
 dummy_Tanaka 
 None 
 None 
 None 
 
 
 10 
 300 
 dummy_Sato 
 None 
 None 
 None 
 
 
 
 
 
 
 id 
 name 
 class 
 age 
 height 
 
 
 
 
 0 
 1 
 Yamada 
 1 
 14 
 150 
 
 
 1 
 9 
 Nagata 
 4 
 11 
 160 
 
 
 2 
 10 
 Sugino 
 2 
 13 
 150 
 
 
 3 
 11 
 Takayama 
 1 
 14 
 155 
 
 
 4 
 12 
 John 
 3 
 16 
 165 
 
 
 Query OK, 0 rows affected.

作成したストアドプロシージャを見たいときは、show create procedureを使います。

In [22]:
show create procedure stpr1;

Procedure 
 sql_mode 
 Create Procedure 
 character_set_client 
 collation_connection 
 Database Collation 
 
 
 
 
 0 
 stpr1 
 NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION 
 CREATE DEFINER=`root`@`%` PROCEDURE `stpr1`()\... 
 utf8 
 utf8_general_ci 
 latin1_swedish_ci

作成したストアドプロシージャを削除したい場合は、dropを使います。

In [23]:
drop procedure stpr1;

Query OK, 0 rows affected.

また、このストアドプロシージャに引数を指定して、実行することも可能です。以下では、int型のxを引数として、設定しています。

In [24]:
create procedure stpr2(in x int)
begin
select * from meibo where age>=x;
select * from meibo2 where age>=x;
end

Query OK, 0 rows affected.

以下を実行すると、meiboとmeibo2のテーブルで、15歳以上のレコードが選択されているのがわかります。

In [25]:
call stpr2(15);

id 
 name 
 class 
 age 
 height 
 
 
 
 
 0 
 4 
 Kato 
 2 
 15 
 150 
 
 
 1 
 6 
 Takeuchi 
 2 
 16 
 155 
 
 
 
 
 
 
 id 
 name 
 class 
 age 
 height 
 
 
 
 
 0 
 12 
 John 
 3 
 16 
 165 
 
 
 Query OK, 0 rows affected.

#### <練習問題 1>

次の処理を実行するストアドプロシージャを作成して実行してみましょう。

・meiboテーブルと同じ構造のテーブルmeibo3を作成

・meiboテーブルのデータをmeibo3テーブルに挿入

・meibo3テーブルにweightカラムを追加

・引数で与えられたBMIとheightカラムの身長から体重を逆算してweightカラムに設定する（※BMI=体重(kg)÷(身長(m)×身長(m)))

・meibo3の内容を出力

### 9.3.2 ストアドファンクション
キーワード：ストアドファンクション

次は、ストアドファンクションです。ストアドプロシージャと同じように作成できますが、値を一つだけ返す点が異なります。ユーザー定義関数とも言われます。作成した関数はcount関数やmax関数のように標準の関数と同じように利用できます。

次は、meiboテーブルにあるデータの平均年齢を返す処理をする関数を作成します。

ストアドプロシージャと同じようにデリミタを変更してから実行してください。

In [26]:
drop function if exists fc1;

Query OK, 0 rows affected.

「create function 関数名 returns 返り値の型」からはじまり、begin~endで記述します。

In [27]:
create function fc1() returns double
begin
declare avgAge double;
select avg(age) into avgAge from meibo;
return avgAge;
end

Query OK, 0 rows affected.

呼び出しは以下のようにします。

In [28]:
select fc1();

fc1() 
 
 
 
 
 0 
 13.5

#### <練習問題 1>

上記の例を参考に、meiboテーブルから平均身長を返す関数を作成して呼んでください。

#### <練習問題 2>

MySQLはナノ秒の取扱ができません。時、分、秒、ナノ秒（小数点以下の秒）の4つの引数を与えると "HH:MM:SS.sssssssss"形式の文字列(varchar(18))を返す関数を作って実行してみましょう。

ヒント：文字列の結合にはconcat関数が利用できます。

***

## 9.4 インデクシングとSQLの高速化
ゴール：データベース上の処理の高速化の手法を知る

### 9.4.1 インデックスの設定
キーワード：インデックス

テーブルからデータを検索するときに、データが膨大にある場合には全件検索は非常に時間がかかります。そこで、テーブルに索引を設定することで、検索がしやすくなり、高速化できます。テーブルに設定する索引を**インデックス**と言います。なお、主キーを設定した場合には、自動的にインデックスも作成されます。ただし、インデックスを設定すれば速くなるというわけではなく、適切に設定することが大事です。

In [29]:
show databases;
use employees;

Database 
 
 
 
 
 0 
 TEST1 
 
 
 1 
 TEST2 
 
 
 2 
 employees 
 
 
 3 
 information_schema 
 
 
 4 
 mysql 
 
 
 5 
 performance_schema 
 
 
 6 
 world 
 
 
 Query OK, 0 rows affected.

In [30]:
show tables;

Tables_in_employees 
 
 
 
 
 0 
 departments 
 
 
 1 
 dept_emp 
 
 
 2 
 dept_manager 
 
 
 3 
 employees 
 
 
 4 
 salaries 
 
 
 5 
 titles

In [31]:
show index from employees;

Table 
 Non_unique 
 Key_name 
 Seq_in_index 
 Column_name 
 Collation 
 Cardinality 
 Sub_part 
 Packed 
 Null 
 Index_type 
 Comment 
 Index_comment 
 
 
 
 
 0 
 employees 
 0 
 PRIMARY 
 1 
 emp_no 
 A 
 299778 
 None 
 None 
 
 BTREE

explainを使うと、クエリがどのような順番でどのような方法でテーブルを検索するかを知ることができます。

In [32]:
explain
select * from employees;

id 
 select_type 
 table 
 type 
 possible_keys 
 key 
 key_len 
 ref 
 rows 
 Extra 
 
 
 
 
 0 
 1 
 SIMPLE 
 employees 
 ALL 
 None 
 None 
 None 
 None 
 299778

条件指定をせずにテーブルの全てを出力しているのでインデックスも利用されていないことがわかります。
9.2.4で実行したクエリをexplainで調べてみましょう。

In [33]:
explain
select 
    c.dept_no
    ,c.dept_name,count(distinct a.emp_no) as cnt
    ,avg(avgSalary) as avgDepSalary 
from dept_emp a
left outer join
(
    select 
        emp_no
        ,avg(salary) as avgSalary 
    from salaries 
group by 1) b
on a.emp_no = b.emp_no 
left outer join departments c
on a.dept_no = c.dept_no
group by 1
having count(*) * 0.5 <= SUM(case when b.avgSalary >= 60000 then 1 else 0 end)

id 
 select_type 
 table 
 type 
 possible_keys 
 key 
 key_len 
 ref 
 rows 
 Extra 
 
 
 
 
 0 
 1 
 PRIMARY 
 a 
 index 
 None 
 emp_no 
 4 
 None 
 331570 
 Using index; Using temporary; Using filesort 
 
 
 1 
 1 
 PRIMARY 
 <derived2> 
 ref 
 key0 
 key0 
 5 
 employees.a.emp_no 
 10 
 
 
 
 2 
 1 
 PRIMARY 
 c 
 eq_ref 
 PRIMARY 
 PRIMARY 
 4 
 employees.a.dept_no 
 1 
 
 
 
 3 
 2 
 DERIVED 
 salaries 
 index 
 None 
 PRIMARY 
 7 
 None 
 2838426

keyに値が入っていればインデックスは利用されています。

In [34]:
select * from employees limit 4;

emp_no 
 birth_date 
 first_name 
 last_name 
 gender 
 hire_date 
 
 
 
 
 0 
 10001 
 1953-09-02 
 Georgi 
 Facello 
 M 
 1986-06-26 
 
 
 1 
 10002 
 1964-06-02 
 Bezalel 
 Simmel 
 F 
 1985-11-21 
 
 
 2 
 10003 
 1959-12-03 
 Parto 
 Bamford 
 M 
 1986-08-28 
 
 
 3 
 10004 
 1954-05-01 
 Chirstian 
 Koblick 
 M 
 1986-12-01

既にあるテーブルにインデックスを追加する場合は以下のように設定します。（以下のコマンドは実行に時間がかかる可能性があります）

In [35]:
create index my_oinde_birth_date on employees(birth_date);

Query OK, 0 rows affected.

確かめると以下のようにインデックスが設定されていることがわかります。

In [36]:
show index from employees;

Table 
 Non_unique 
 Key_name 
 Seq_in_index 
 Column_name 
 Collation 
 Cardinality 
 Sub_part 
 Packed 
 Null 
 Index_type 
 Comment 
 Index_comment 
 
 
 
 
 0 
 employees 
 0 
 PRIMARY 
 1 
 emp_no 
 A 
 299778 
 None 
 None 
 
 BTREE 
 
 
 
 
 1 
 employees 
 1 
 my_oinde_birth_date 
 1 
 birth_date 
 A 
 9670 
 None 
 None 
 
 BTREE

インデックスは以下のように削除します。

In [37]:
drop index my_oinde_birth_date on employees;

Query OK, 0 rows affected.

確かめると以下のようにインデックスが削除されていることがわかります。

In [38]:
show index from employees;

Table 
 Non_unique 
 Key_name 
 Seq_in_index 
 Column_name 
 Collation 
 Cardinality 
 Sub_part 
 Packed 
 Null 
 Index_type 
 Comment 
 Index_comment 
 
 
 
 
 0 
 employees 
 0 
 PRIMARY 
 1 
 emp_no 
 A 
 299778 
 None 
 None 
 
 BTREE

大量のデータから一部を検索するときはインデックスが利用されることが重要になります。
一回のwhere句やjoin、order by、group by の条件指定では一個のインデックスしか使えません。複数のカラムを条件指定しないと検索結果を絞り込めない場合は複合インデックスを指定します。salariesテーブルは元々、主キーがemp_noとfrom_dateの複合インデックスになっていますが、from_date、to_dateを同時に用いて検索したい場合は次のようにインデックスを設定すると良いでしょう。

In [39]:
create index from_to_date on salaries(from_date,to_date);

Query OK, 0 rows affected.

作成されたインデックスを確認します。

In [40]:
show index from salaries;

Table 
 Non_unique 
 Key_name 
 Seq_in_index 
 Column_name 
 Collation 
 Cardinality 
 Sub_part 
 Packed 
 Null 
 Index_type 
 Comment 
 Index_comment 
 
 
 
 
 0 
 salaries 
 0 
 PRIMARY 
 1 
 emp_no 
 A 
 709606 
 None 
 None 
 
 BTREE 
 
 
 
 
 1 
 salaries 
 0 
 PRIMARY 
 2 
 from_date 
 A 
 2838426 
 None 
 None 
 
 BTREE 
 
 
 
 
 2 
 salaries 
 1 
 emp_no 
 1 
 emp_no 
 A 
 709606 
 None 
 None 
 
 BTREE 
 
 
 
 
 3 
 salaries 
 1 
 from_to_date 
 1 
 from_date 
 A 
 11976 
 None 
 None 
 
 BTREE 
 
 
 
 
 4 
 salaries 
 1 
 from_to_date 
 2 
 to_date 
 A 
 135163 
 None 
 None 
 
 BTREE

インデックスを追加すると、検索は高速になりますが、インデックスの作成に時間がかかるためレコードの追加が遅くなったり、インデックスの情報を格納するためにディスク容量も必要になるというデメリットがあります。デメリットを把握した上で設定しましょう。

レコード数が少なかったり、ランダムアクセスが高速なSSDを利用している場合は必ずしもインデックスが必要ではないため、設定していても利用されない場合があります。むやみにインデックスを設定することは避けましょう。

また、インデックスで絞り込んでもテーブルの大部分のレコードが残るような場合、インデックスを介して検索するよりインデックスを利用しないほうが高速な場合もあります。

#### <練習問題 1>

dept_empテーブルのfrom_date、to_dateに複合インデックスを設定してみましょう。

### 9.4.2 インデックスの種類
キーワード：Bツリーインデックス、ハッシュインデックス、Rツリーインデックス

参考ですが、インデックスのデータの型の種類とそれぞれの特徴について以下、簡単に説明します。興味のある方や参考文献などを見て調べてみてください。

- **Bツリー(balanced tree)インデックス**  
一番人気が高く、インデックスの型として最もよく使われます。  
柔軟性、サイズ、全体的な性能のバランスをうまくとっています。


- **ハッシュインデックス**  
二番目に人気が高いです。  
ツリーよりもハッシュテーブルに似ていて、各キーにハッシュ関数を実行した結果に基づいて並べます。   


- **Rツリーインデックス**  
空間データまたはN次元データを扱うときに用いられます。  
地図や地球科学のアプリケーションで幅広く使われます。  

### 9.4.3 SQLのパフォーマンスチューニング
キーワード：特になし

インデックスを使用する際は、高速化のため、条件式の左辺はカラムのみを入れましょう。以下は比較のため、emp_noを係数倍して、それが400000より大きいものを抽出するクエリを示してあります。

**注意：現在の環境ではクエリが返ってくるまでの実行時間は表示されませんが、terminal上で実行すると見ることができますので、参考にしてください（なお、terminalで「mysql -h"zdb" -P"3306" -u root -p"gci"」のコマンドを実行すると、mariadbでsqlがたたけるようになります。）**

In [41]:
select * from employees
where emp_no * 0.80001 > 400000;

emp_no 
 birth_date 
 first_name 
 last_name 
 gender 
 hire_date 
 
 
 
 
 0 
 499994 
 1952-02-26 
 Navin 
 Argence 
 F 
 1990-04-24 
 
 
 1 
 499995 
 1958-09-24 
 Dekang 
 Lichtner 
 F 
 1993-01-12 
 
 
 2 
 499996 
 1953-03-07 
 Zito 
 Baaz 
 M 
 1990-09-27 
 
 
 3 
 499997 
 1961-08-03 
 Berhard 
 Lenart 
 M 
 1986-04-21 
 
 
 4 
 499998 
 1956-09-05 
 Patricia 
 Breugel 
 M 
 1993-10-13 
 
 
 5 
 499999 
 1958-05-01 
 Sachin 
 Tsukuda 
 M 
 1997-11-30

左辺に演算を入れずに、そのカラムのままにしておく設定だと、クエリが返ってくる時間が短くなります。上記の結果と比較してみてください。

In [42]:
select * from employees
where emp_no > 400000/0.80001;

emp_no 
 birth_date 
 first_name 
 last_name 
 gender 
 hire_date 
 
 
 
 
 0 
 499994 
 1952-02-26 
 Navin 
 Argence 
 F 
 1990-04-24 
 
 
 1 
 499995 
 1958-09-24 
 Dekang 
 Lichtner 
 F 
 1993-01-12 
 
 
 2 
 499996 
 1953-03-07 
 Zito 
 Baaz 
 M 
 1990-09-27 
 
 
 3 
 499997 
 1961-08-03 
 Berhard 
 Lenart 
 M 
 1986-04-21 
 
 
 4 
 499998 
 1956-09-05 
 Patricia 
 Breugel 
 M 
 1993-10-13 
 
 
 5 
 499999 
 1958-05-01 
 Sachin 
 Tsukuda 
 M 
 1997-11-30

他、無駄な中間テーブルは減らして、having句を活用しましょう。以下は比較のための中間テーブルを利用したクエリです

In [43]:
select * from
(select emp_no,max(salary) as maxSalary from salaries group by 1) tmp
where maxSalary > 155000;

emp_no 
 maxSalary 
 
 
 
 
 0 
 43624 
 158220 
 
 
 1 
 47978 
 155709 
 
 
 2 
 109334 
 155377 
 
 
 3 
 253939 
 155513 
 
 
 4 
 254466 
 156286

若干ですが、クエリが返ってくる時間が短くなります。

In [44]:
select 
    emp_no
    ,max(salary) as maxSalary
from salaries 
group by 1
having maxSalary > 155000;

emp_no 
 maxSalary 
 
 
 
 
 0 
 43624 
 158220 
 
 
 1 
 47978 
 155709 
 
 
 2 
 109334 
 155377 
 
 
 3 
 253939 
 155513 
 
 
 4 
 254466 
 156286

演算処理はレコード件数を絞った後に実行されるようにしましょう。

9.2.3節で実行したクエリを高速化してみます。

(**ここから先の結果は、環境によっては、特に変化がなかったり、遅かったりする可能性もあるので、参考に見ておいてください**)

explainを実行してみましょう。

In [45]:
explain
select 
    x0.emp_no
    ,case when x1.emp_no is not null then 'OK' else '' end as "Staff"
    ,case when x2.emp_no is not null then 'OK' else '' end as "Senior Staff"
    ,case when x3.emp_no is not null then 'OK' else '' end as "Engineer"
from (select distinct emp_no from employees ) x0
left outer join (select emp_no from titles where title ="Staff") x1 on x0.emp_no = x1.emp_no
left outer join (select emp_no from titles where title ="Senior Staff") x2 on x0.emp_no = x2.emp_no
left outer join (select emp_no from titles where title ="Engineer") x3 on x0.emp_no = x3.emp_no
limit 10;

id 
 select_type 
 table 
 type 
 possible_keys 
 key 
 key_len 
 ref 
 rows 
 Extra 
 
 
 
 
 0 
 1 
 PRIMARY 
 <derived2> 
 ALL 
 None 
 None 
 None 
 None 
 299778 
 
 
 
 1 
 1 
 PRIMARY 
 titles 
 ref 
 PRIMARY,emp_no 
 PRIMARY 
 56 
 x0.emp_no,const 
 1 
 Using where; Using index 
 
 
 2 
 1 
 PRIMARY 
 titles 
 ref 
 PRIMARY,emp_no 
 PRIMARY 
 56 
 x0.emp_no,const 
 1 
 Using where; Using index 
 
 
 3 
 1 
 PRIMARY 
 titles 
 ref 
 PRIMARY,emp_no 
 PRIMARY 
 56 
 x0.emp_no,const 
 1 
 Using where; Using index 
 
 
 4 
 2 
 DERIVED 
 employees 
 index 
 None 
 PRIMARY 
 4 
 None 
 299778 
 Using index

このクエリはx0の部分が主になって、x1,x2,x3が外部結合されています。

explainの結果、一番最初に選択されているテーブル(x0)の行数が291,937となっています。これは、limit句が一番外にあるためです。joinやcase文の演算を実行してからレコードがlimit句で10行だけ抽出されています。x0の段階でlimit 10を指定しておくとその他の処理も10行分しか実行しないため高速になります。

x0のサブクエリの時点でlimit句を指定してexplainを実行してみましょう

In [46]:
explain
select 
    x0.emp_no
    ,case when x1.emp_no is not null then 'OK' else '' end as "Staff"
    ,case when x2.emp_no is not null then 'OK' else '' end as "Senior Staff"
    ,case when x3.emp_no is not null then 'OK' else '' end as "Engineer"
from (select distinct emp_no from employees limit 10) x0
left outer join (select emp_no from titles where title ="Staff") x1 on x0.emp_no = x1.emp_no
left outer join (select emp_no from titles where title ="Senior Staff") x2 on x0.emp_no = x2.emp_no
left outer join (select emp_no from titles where title ="Engineer") x3 on x0.emp_no = x3.emp_no;

id 
 select_type 
 table 
 type 
 possible_keys 
 key 
 key_len 
 ref 
 rows 
 Extra 
 
 
 
 
 0 
 1 
 PRIMARY 
 <derived2> 
 ALL 
 None 
 None 
 None 
 None 
 10 
 
 
 
 1 
 1 
 PRIMARY 
 titles 
 ref 
 PRIMARY,emp_no 
 PRIMARY 
 56 
 x0.emp_no,const 
 1 
 Using where; Using index 
 
 
 2 
 1 
 PRIMARY 
 titles 
 ref 
 PRIMARY,emp_no 
 PRIMARY 
 56 
 x0.emp_no,const 
 1 
 Using where; Using index 
 
 
 3 
 1 
 PRIMARY 
 titles 
 ref 
 PRIMARY,emp_no 
 PRIMARY 
 56 
 x0.emp_no,const 
 1 
 Using where; Using index 
 
 
 4 
 2 
 DERIVED 
 employees 
 index 
 None 
 PRIMARY 
 4 
 None 
 299778 
 Using index

一番最初に選択されるテーブル(x0)の行数が10になりました。

実行してみます。

In [47]:
select 
    x0.emp_no
    ,case when x1.emp_no is not null then 'OK' else '' end as "Staff"
    ,case when x2.emp_no is not null then 'OK' else '' end as "Senior Staff"
    ,case when x3.emp_no is not null then 'OK' else '' end as "Engineer"
from (select distinct emp_no from employees limit 10) x0
left outer join (select emp_no from titles where title ="Staff") x1 on x0.emp_no = x1.emp_no
left outer join (select emp_no from titles where title ="Senior Staff") x2 on x0.emp_no = x2.emp_no
left outer join (select emp_no from titles where title ="Engineer") x3 on x0.emp_no = x3.emp_no;

emp_no 
 Staff 
 Senior Staff 
 Engineer 
 
 
 
 
 0 
 10001 
 
 
 
 
 
 1 
 10002 
 OK 
 
 
 
 
 2 
 10003 
 
 
 
 
 
 3 
 10004 
 
 
 OK 
 
 
 4 
 10005 
 OK 
 OK 
 
 
 
 5 
 10006 
 
 
 
 
 
 6 
 10007 
 OK 
 OK 
 
 
 
 7 
 10008 
 
 
 
 
 
 8 
 10009 
 
 
 OK 
 
 
 9 
 10010 
 
 
 OK

実行時間が大幅削減されました。

さらに、x0テーブルはemp_noの重複を排除するためにdistinctを指定していますが、employeesテーブルの主キーなので重複はありません。x0のサブクエリは廃止できます。

In [48]:
select 
    x0.emp_no
    ,case when x1.emp_no is not null then 'OK' else '' end as "Staff"
    ,case when x2.emp_no is not null then 'OK' else '' end as "Senior Staff"
    ,case when x3.emp_no is not null then 'OK' else '' end as "Engineer"
from employees  x0
left outer join (select emp_no from titles where title ="Staff") x1 on x0.emp_no = x1.emp_no
left outer join (select emp_no from titles where title ="Senior Staff") x2 on x0.emp_no = x2.emp_no
left outer join (select emp_no from titles where title ="Engineer") x3 on x0.emp_no = x3.emp_no
limit 10;

emp_no 
 Staff 
 Senior Staff 
 Engineer 
 
 
 
 
 0 
 10001 
 
 
 
 
 
 1 
 10002 
 OK 
 
 
 
 
 2 
 10003 
 
 
 
 
 
 3 
 10004 
 
 
 OK 
 
 
 4 
 10005 
 OK 
 OK 
 
 
 
 5 
 10006 
 
 
 
 
 
 6 
 10007 
 OK 
 OK 
 
 
 
 7 
 10008 
 
 
 
 
 
 8 
 10009 
 
 
 OK 
 
 
 9 
 10010 
 
 
 OK

この他、SQLの高速化には以下のような手法があります。詳細の説明は省略しますが、DBMSの公式サイトや書籍等を参照してください。

##### SQL実行時のプロファイリング結果を利用する。

プロファイリング機能を利用して、クエリの実行時間のうちどれだけの時間をどの処理で使用したかを確認しつつ、時間のかかっている原因を取り除きます。explainの結果とあわせて利用します。

##### パーティショニングを利用する。

1つのテーブルを複数に分割することで処理の高速化と格納効率を向上することが見込めます。テーブル定義にパーティショニングを設定することでテーブルを分割して高速になる場合があります。ただし、分割したテーブルを全て参照するようなSQLになる場合は逆に遅くなる場合もありますので注意してください。

##### DBMSの設定値をチューニングする。

DBMSの確保するメモリが足りないために一時テーブルをファイルに出力して処理すると急激に遅くなります。DBMSの確保するキャッシュ容量を調整することで回避できる場合があります。その他、DBMSの設定値を変更することでパフォーマンスを向上させることができる場合があります。

***

## 9.5 その他
ゴール：ファイルのやり取りやデータモデリング、正規化の概念を知る

### 9.5.1 データベースとファイルのやり取り
キーワード：データのインポート、エクスポート

データのファイル形式は様々あり、ここではよく扱われるCSV(Comma Separated Values)ファイルをデータベースに取り込むインポートや、データをファイルに書き出すエクスポート等について簡単に紹介します。

まずは、インポートです。これまでサンプルのデータを準備するときに、手で一部入力をして、実行していました。もちろん、大量のデータがある場合に、手で入力するのは現実的ではなく、CSVファイルにあるデータを自動的に入れる方法があり、「load data infile 'ファイル名' into table テーブル名 オプション;」のような記述で実行します。
ファイル名は、そのディレクトリ名を含むこともできます。オプションは、データの区切り文字や先頭をスキップする時などに使います。
一方、エクスポートについては、「select * into outfile 'ファイル名' オプション from テーブル名」のように記述をして、ファイルをアウトプットします。もし、大量のファイルをデータベースに入れたい場合や、その逆をすることがあれば、これらの方法を検討してみてください。

この他、SQLが記述されたファイルを実行する「source ファイル名」や、データベースをバックアップする方法（ダンプ）や、それを再び戻す方法（リストア）もあります。詳しくはMySQLの公式サイトを参照してみてください。

データのインポート：https://dev.mysql.com/doc/refman/5.6/ja/load-data.html

データのエクスポート：https://dev.mysql.com/doc/refman/5.6/ja/select-into.html

データベースのバックアップ・リストア：https://dev.mysql.com/doc/refman/5.6/ja/mysqldump.html

### 9.5.2 データモデリングと正規化
キーワード：データモデリング、正規化

前章では、データベースにテーブルがない場合は、テーブルを作成してきました。顧客の名前を管理する必要があれば、顧客名のカラムを作成し、それだけでは名前に重複がある可能性もあるので、一意に定まるIDのカラムを設定したり、他に属性情報等があれば、それらのカラムを作成します。これらのテーブル設計の要件は、ビジネスによって異なり、はじめにしっかりとやらないと、後で困ること（カラムが足りない、後から追加したいなどのニーズが出てきたときに柔軟に変更できるかどうかなど）も出てきます。このように、必要なデータ項目を洗い出して、テーブル設計していくことを**データモデリング**といいます。

このデータモデリングを進めていく上で、重要なことが**正規化**です。正規化の目的は、効率よくデータを扱うために整理することにあります。具体的には、あるテーブルの列において、データが繰り返し出てきて、別のテーブルにしたほうが、更新などもしやすいというケースがあります。この正規化には第1正規形から第5正規形、他にも色々とありますので、興味ある方は以下の参考文献やネットなどで調べてみてください。

>[参考文献]

>『Webエンジニアのための データベース技術[実践]入門 (Software Design plus)』（松信 嘉範 (著)、技術評論社）

以上で、この章は完了です。お疲れ様でした。再掲ですが、以下の参考文献などを読んで、是非中級者への道に進んでください。

>[参考文献]

>『SQL実践入門──高速でわかりやすいクエリの書き方 (WEB+DB PRESS plus) 』（ミック  (著)、技術評論社）

>『プログラマのためのSQL 第4版』（ジョー・セルコ (著), Joe Celko (著), ミック (監修, 翻訳)、翔泳社）

>『達人に学ぶ SQL徹底指南書』（ミック  (著)、翔泳社）

>『達人に学ぶDB設計 徹底指南書 初級者で終わりたくないあなたへ 』（ミック  (著)、翔泳社）

>『ビッグデータ分析・活用のためのSQLレシピ』（加嵜 長門  (著), 田宮 直人  (著), 丸山 弘詩 (編集)、マイナビ出版 ）

***

## 9.6 総合問題

### 9.6.1 総合問題1

以下のデータベースを使って、次の問題を考えて実行してみましょう。（注意事項等を読んでください。）

https://dev.mysql.com/doc/sakila/en/

ターミナル(もしくは画面上にあるKernelをPythonに変更して)から以下のコマンドを入力します。

-------------------
wget http://downloads.mysql.com/docs/sakila-db.tar.gz  
tar zxf sakila-db.tar.gz  
cd sakila-db  
mysql -h"zdb" -P"3306" -u root -p"gci" < sakila-schema.sql  
mysql -h"zdb" -P"3306" -u root -p"gci" < sakila-data.sql  

-------------------

(1) 支払金額の総額が100ドル以上の顧客名と支払総額を表示してください。

(2) (1)の結果を新しいテーブルに格納するストアドプロシージャを作成してください。ただし、支払総額の閾値は引数で指定できるようにしてください。

### 9.6.2 総合問題2

以下のデータベースを使って 後の問に答えてください。（こちらも注意事項等を読んでください。）

https://sites.google.com/a/dlpage.phi-integration.com/pentaho/mondrian/mysql-foodmart-database/foodmart_mysql.tar.gz?attredirects=0

参考の解説等はこちらになります。

https://github.com/julianhyde/foodmart-data-hsqldb

http://www.fromdual.ch/foodmart-old

**まずはデータを入れるために、データベースfoodmartを作成してください。**

次に、ターミナル(もしくは画面上にあるKernelをPythonに変更して)から以下のコマンドを入力します。

-------------------
mkdir foodmart  
cd foodmart  
wget http://pentaho.dlpage.phi-integration.com/mondrian/mysql-foodmart-database/foodmart_mysql.tar.gz  
tar zxf  foodmart_mysql.tar.gz  
mysql -h"zdb" -P"3306" -u root -p"gci" foodmart < foodmart_mysql.sql  

-------------------

(1) 1998年1月の従業員の給料ランキングを10位まで表示してください。

(2) sales_fact_1998テーブルの中で、同じ人が同じ時間に買っている商品の2つの組み合わせの中で、購買回数が多い2つの商品はどれとどれでしょうか。TOP5を抽出して、商品名(product_name)でそれぞれ答えてください。なお、商品はproduct_idベースでカウントして、同じ人とはcustomer_idが同じことを意味し、同じ時間とはtime_idが同じことを意味します。なお、これをバスケット分析（同時併売）といい、期間併売で計算することもあります。